File: vane-portals/src/main/java/org/oddlama/vane/portals/PortalActivator.java
package org.oddlama.vane.portals;

import static org.oddlama.vane.util.PlayerUtil.swingArm;

import org.bukkit.Material;
import org.bukkit.Tag;
import org.bukkit.block.BlockFace;
import org.bukkit.block.data.type.Repeater;
import org.bukkit.block.data.type.Switch;
import org.bukkit.event.Event;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.BlockRedstoneEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.oddlama.vane.core.Listener;
import org.oddlama.vane.core.module.Context;
import org.oddlama.vane.portals.portal.PortalBlock;

public class PortalActivator extends Listener<Portals> {

    public PortalActivator(Context<Portals> context) {
        super(context);
    }

    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = false)
    public void onPlayerInteractConsole(final PlayerInteractEvent event) {
        if (!event.hasBlock() || event.getAction() != Action.RIGHT_CLICK_BLOCK) {
            return;
        }

        if (event.useInteractedBlock() == Event.Result.DENY) {
            return;
        }

        // Abort if the table is not a console
        final var block = event.getClickedBlock();
        final var portalBlock = getModule().portalBlockFor(block);
        if (portalBlock == null || portalBlock.type() != PortalBlock.Type.CONSOLE) {
            return;
        }

        event.setUseInteractedBlock(Event.Result.DENY);
        event.setUseItemInHand(Event.Result.DENY);

        final var player = event.getPlayer();
        final var portal = getModule().portalFor(portalBlock);
        if (portal.openConsole(getModule(), player, block)) {
            swingArm(player, event.getHand());
        }
    }

    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = true)
    public void onPlayerInteractSwitch(final PlayerInteractEvent event) {
        if (!event.hasBlock() || event.getAction() != Action.RIGHT_CLICK_BLOCK) {
            return;
        }

        if (event.useInteractedBlock() == Event.Result.DENY) {
            return;
        }

        final var block = event.getClickedBlock();
        final boolean allowDisable;
        if (block.getType() == Material.LEVER) {
            allowDisable = true;
        } else if (Tag.BUTTONS.isTagged(block.getType())) {
            allowDisable = false;
        } else {
            return;
        }

        // Get base block the switch is attached to
        final var bswitch = (Switch) block.getBlockData();
        final BlockFace attachedFace;
        switch (bswitch.getAttachedFace()) {
            default:
            case WALL:
                attachedFace = bswitch.getFacing().getOppositeFace();
                break;
            case CEILING:
                attachedFace = BlockFace.UP;
                break;
            case FLOOR:
                attachedFace = BlockFace.DOWN;
                break;
        }

        // Find controlled portal
        final var base = block.getRelative(attachedFace);
        final var portal = getModule().controlledPortal(base);
        if (portal == null) {
            return;
        }

        final var player = event.getPlayer();
        final var active = getModule().isActivated(portal);
        if (bswitch.isPowered() && allowDisable) {
            if (!active) {
                return;
            }

            // Switch is being switched off → deactivate
            if (!portal.deactivate(getModule(), player)) {
                event.setUseInteractedBlock(Event.Result.DENY);
                event.setUseItemInHand(Event.Result.DENY);
            }
        } else {
            if (active) {
                return;
            }

            // Switch is being switched on → activate
            if (!portal.activate(getModule(), player)) {
                event.setUseInteractedBlock(Event.Result.DENY);
                event.setUseItemInHand(Event.Result.DENY);
            }
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onBlockRedstone(final BlockRedstoneEvent event) {
        // Only on rising edge.
        if (event.getOldCurrent() != 0 || event.getNewCurrent() == 0) {
            return;
        }

        // Only repeaters
        final var block = event.getBlock();
        if (block.getType() != Material.REPEATER) {
            return;
        }

        // Get the block it's pointing towards. (Opposite of block's facing for repeaters)
        final var repeater = (Repeater) block.getBlockData();
        final var intoBlock = block.getRelative(repeater.getFacing().getOppositeFace());

        // Find controlled portal
        final var portal = getModule().portalFor(intoBlock);
        if (portal == null) {
            return;
        }

        portal.activate(getModule(), null);
    }
}

File: vane-portals/src/main/kotlin/org/oddlama/vane/portals/PortalActivator.kt
package org.oddlama.vane.portals

import org.bukkit.Material
import org.bukkit.Tag
import org.bukkit.block.BlockFace
import org.bukkit.block.data.FaceAttachable.AttachedFace
import org.bukkit.block.data.type.Repeater
import org.bukkit.block.data.type.Switch
import org.bukkit.event.Event
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockRedstoneEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.oddlama.vane.core.Listener
import org.oddlama.vane.core.module.Context
import org.oddlama.vane.portals.portal.Portal
import org.oddlama.vane.portals.portal.PortalBlock
import org.oddlama.vane.portals.portal.PortalBlockLookup
import org.oddlama.vane.util.PlayerUtil

/** Handles player and redstone interactions that activate portal behavior. */
class PortalActivator(context: Context<Portals?>?) : Listener<Portals?>(context) {
    /**
     * Helper that validates a PlayerInteractEvent is a right-click on a block the module
     * should handle and returns a pair of the resolved module and clicked block.
     * Returns null if any precondition fails.
     */
    private fun requireModuleAndClickedBlock(event: PlayerInteractEvent): Pair<Portals, org.bukkit.block.Block>? {
        val module = module ?: return null

        if (!event.hasBlock() || event.action != Action.RIGHT_CLICK_BLOCK) {
            return null
        }

        if (event.useInteractedBlock() == Event.Result.DENY) {
            return null
        }

        val block = event.clickedBlock ?: return null

        return Pair(module, block)
    }
    /** Opens the portal console menu when a portal console block is used. */
    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = false)
    fun onPlayerInteractConsole(event: PlayerInteractEvent) {
        val (module, block) = requireModuleAndClickedBlock(event) ?: return
        val portalBlock: PortalBlockLookup? = module.portalBlockFor(block)
        if (portalBlock == null || portalBlock.type() != PortalBlock.Type.CONSOLE) {
            return
        }

        event.setUseInteractedBlock(Event.Result.DENY)
        event.setUseItemInHand(Event.Result.DENY)

        val player = event.player
        val portal: Portal = module.portalFor(portalBlock)
        if (portal.openConsole(module, player, block)) {
            PlayerUtil.swingArm(player, event.hand!!)
        }
    }

    /** Toggles controlled portals when a bound switch is interacted with. */
    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = true)
    fun onPlayerInteractSwitch(event: PlayerInteractEvent) {
        val (module, block) = requireModuleAndClickedBlock(event) ?: return
        val allowDisable = when {
            block.type == Material.LEVER -> true
            Tag.BUTTONS.isTagged(block.type) -> false
            else -> return
        }

        // Get base block the switch is attached to
        val bswitch = block.blockData as Switch
        val attachedFace: BlockFace = when (bswitch.attachedFace) {
            AttachedFace.WALL -> bswitch.facing.getOppositeFace()
            AttachedFace.CEILING -> BlockFace.UP
            AttachedFace.FLOOR -> BlockFace.DOWN
        }

        // Find controlled portal
        val base = block.getRelative(attachedFace)
        val portal = module.controlledPortal(base) ?: return

        val player = event.player
        val active = module.isActivated(portal)
        if (bswitch.isPowered && allowDisable) {
            if (!active) {
                return
            }

            // Switch is being switched off → deactivate
            if (!portal.deactivate(module, player)) {
                event.setUseInteractedBlock(Event.Result.DENY)
                event.setUseItemInHand(Event.Result.DENY)
            }
        } else {
            if (active) {
                return
            }

            // Switch is being switched on → activate
            if (!portal.activate(module, player)) {
                event.setUseInteractedBlock(Event.Result.DENY)
                event.setUseItemInHand(Event.Result.DENY)
            }
        }
    }

    /** Activates portals when a repeater powers their gateway block. */
    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onBlockRedstone(event: BlockRedstoneEvent) {
        val module = module ?: return

        // Only on rising edge.
        if (event.oldCurrent != 0 || event.newCurrent == 0) {
            return
        }

        // Only repeaters
        val block = event.getBlock()
        if (block.type != Material.REPEATER) {
            return
        }

        // Get the block it's pointing towards. (Opposite of block's facing for repeaters)
        val repeater = block.blockData as Repeater
        val intoBlock = block.getRelative(repeater.facing.getOppositeFace())

        // Find controlled portal
        val portal = module.portalFor(intoBlock) ?: return

        portal.activate(module, null)
    }
}
